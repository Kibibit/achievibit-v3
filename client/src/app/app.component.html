<header>
  <nav style="gap: 0.5em; --outline-variant: transparent;">
    <!-- <button class="circle transparent">
      <i>menu</i>
    </button> -->
    <h5>
      <a class="logo" [routerLink]="['/']">
        <img src="https://kibibit.io/kibibit-assets/logo-achi.svg" class="logo" alt="achievibit logo">
        achievibit
      </a>
    </h5>
    <a class="chip" routerLinkActive="fill" [routerLink]="['/repos']">Repositories</a>
    <a class="chip" routerLinkActive="fill" [routerLink]="['/users']">Users</a>
    <a class="chip" routerLinkActive="fill" [routerLink]="['/orgs']">Organizations</a>
    <div class="max"></div>
    <button *ngIf="loggedInUser" class="circle transparent" type="button" (click)="openMenu()">
      <img class="responsive" alt="user avatar" [src]="loggedInUser.avatar">
      <!-- <dialog *ngIf="menuOpen" #userMenuDialog class="right blur" [class.active]="menuOpen">
        <article class="transparent" style="padding: 0em;">
          <div class="row">
            <img class="circle small" alt="user avatar" [src]="loggedInUser.avatar">
            <div class="max">
              <div style="font-size: 1.1em"><strong>{{ loggedInUser.username }}</strong></div>
              <div style="font-size: 0.8em">Some text here</div>
            </div>
            <button class="circle small transparent" type="button" (click)="closeMenu()">
              <i>close</i>
            </button>
          </div>
        </article>
        <nav class="drawer">
          <a class="row" [routerLink]="['/profile']">
            <i class="material-symbols-outlined">
              person
            </i>
            <span>Your Profile</span>
          </a>
          <a class="row" [routerLink]="['/profile/integrations']">
            <i class="material-symbols-outlined">
              integration_instructions
            </i>
            <span>Integrations</span>
          </a>
          <a class="row" [routerLink]="['/profile/settings']">
            <i class="material-symbols-outlined">
              settings
            </i>
            <span>Settings</span>
          </a>
          <hr class="small">
          <a class="row" [routerLink]="['/app-status']">
            <i class="material-symbols-outlined">
              heart_check
            </i>
            <span>Achievibit Status</span>
          </a>
          <hr class="small">
          <a class="row" tabindex="0" (click)="logout()" (keydown.enter)="logout()">
            <i class="material-symbols-outlined">
              logout
            </i>
            <span>Sign out</span>
          </a>
        </nav>
      </dialog> -->
      <menu *ngIf="menuOpen" class="left no-wrap">
        <a class="row" routerLinkActive="fill" [routerLink]="['/profile/overview']">
          <i class="material-symbols-outlined">
            person
          </i>
          <span>Your Profile</span>
        </a>
        <a class="row" routerLinkActive="fill" [routerLink]="['/profile/integrations']">
          <i class="material-symbols-outlined">
            integration_instructions
          </i>
          <span>Integrations</span>
        </a>
        <a class="row" routerLinkActive="fill" [routerLink]="['/profile/settings']">
          <i class="material-symbols-outlined">
            settings
          </i>
          <span>Settings</span>
        </a>
        <hr class="small">
        <a class="row" routerLinkActive="fill" [routerLink]="['/app-status']">
          <i class="material-symbols-outlined">
            heart_check
          </i>
          <span>Achievibit Status</span>
        </a>
        <hr class="small">
        <a class="row" tabindex="0" (click)="logout()" (keydown.enter)="logout()">
          <i class="material-symbols-outlined">
            logout
          </i>
          <span>Sign out</span>
        </a>
      </menu>
    </button>
  </nav>
</header>
<progress *ngIf="loading$ | async" class="orange-text"></progress>
<main class="responsive padding">
  <router-outlet />
</main>
<kb-achievement *ngIf="loggedInUser" [username]="loggedInUser.username"></kb-achievement>