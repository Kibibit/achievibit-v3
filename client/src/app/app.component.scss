main {
  position: relative;
  padding: 0;
}

// nav a.active {
//   color: var(--primary);
// }