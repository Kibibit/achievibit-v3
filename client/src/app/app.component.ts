import { StatusCodes } from 'http-status-codes';
import { NgIf } from '@angular/common';
import { Component, ElementRef, OnInit, ViewChild } from '@angular/core';
import { Router, RouterOutlet } from '@angular/router';

import { User } from '@kibibit/achievibit-sdk';

import { MeApiService } from './services/api/me.service';
import { AuthService } from './services/auth.service';
import { LoaderService } from './services/loader.service';
import { SocketService } from './services/socket.service';
import { AchievementComponent } from './shared/achievement/achievement.component';
import { HeaderComponent } from './shared/header/header.component';
import { MiniGameComponent } from './shared/mini-game/mini-game.component';

@Component({
  selector: 'kb-root',
  standalone: true,
  imports: [
    RouterOutlet,
    NgIf,
    AchievementComponent,
    AchievementComponent,
    HeaderComponent,
    MiniGameComponent
  ],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss'
})
export class AppComponent implements OnInit {
  hideUserState: boolean = true;
  loading$ = this.loaderService.loading$;
  loggedInUser?: User;
  menuOpen = false;
  messages: string[] = [];
  newMessage: string = '';

  @ViewChild('userMenuDialog') userMenuDialog!: ElementRef<HTMLDialogElement>;

  constructor(
    private socketService: SocketService,
    private meApiService: MeApiService,
    private router: Router,
    private loaderService: LoaderService,
    private readonly authService: AuthService
  ) {
    (console as any).image = function(url: string, size = 100) {
      const image = new Image();
      image.src = url;
      image.onload = function() {
        const style = [
          'font-size: 1px;',
          'padding: ' + (this as any).height / 100 * size + 'px ' + (this as any).width / 100 * size + 'px;',
          'background: url(' + url + ') no-repeat;',
          'background-size: contain;'
        ].join(' ');
        console.log('%c ', style);
      };
    };
    const imageUrl = 'https://kibibit.io/kibibit-assets/1x/long-white.png';

    this.imageUrlToBase64(imageUrl)
      .then((base64) => {
        (console as any).image(base64, 30);
      })
      .catch(console.error);
  }

  private imageUrlToBase64(url: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const img = new Image();
      // Important for cross-origin images!
      img.crossOrigin = 'Anonymous';
      img.onload = function() {
        const canvas = document.createElement('canvas');
        canvas.width = img.width;
        canvas.height = img.height;

        const ctx = canvas.getContext('2d');
        if (!ctx) return reject(new Error('Could not get canvas context'));

        ctx.drawImage(img, 0, 0);
        const dataURL = canvas.toDataURL('image/png');
        resolve(dataURL);
      };
      img.onerror = reject;
      img.src = url;
    });
  }

  ngOnInit(): void {
    // do I even need this here? if I move the navbar element into a component,
    // I can just hide it there and only redirect to login if the user tries to
    // access a page that requires authentication
    this
      .meApiService
      .checkUserLoggedIn()
      .subscribe({
        next: (user) => {
          this.authService.setUser(user);
          this.loggedInUser = user;
          this.hideUserState = false;
        },
        error: (error: { status: number; message: string }) => {
          this.hideUserState = false;
          if (error.status === StatusCodes.UNAUTHORIZED) {
            console.log('User not logged in:');
            return;
          }

          // Handle other errors
          // window.location.href = '/login';
        }
      });
  }
}
