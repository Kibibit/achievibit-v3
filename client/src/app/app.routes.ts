import { Routes } from '@angular/router';

import { HomeComponent } from './home/home.component';
import { KeyboardInputComponent } from './keyboard-input/keyboard-input.component';
import { UserProfileComponent } from './user-profile/user-profile.component';

export const routes: Routes = [
  { path: 'keyboard', component: KeyboardInputComponent },
  { path: 'profile', component: UserProfileComponent },
  { path: 'home', component: HomeComponent },
  { path: '', redirectTo: 'home', pathMatch: 'full' }
  // { path: 'first-component', component: FirstComponent }
];
