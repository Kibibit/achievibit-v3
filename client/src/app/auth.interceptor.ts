import { catchError, tap } from 'rxjs';
import { DOCUMENT } from '@angular/common';
import { HttpErrorResponse, HttpInterceptorFn } from '@angular/common/http';
import { inject } from '@angular/core';

export const authInterceptor: HttpInterceptorFn = (req, next) => {
  const document: Document = inject(DOCUMENT);
  return next(req).pipe(
    tap((httpEvent) => {
      // Checks if the HttpEvent is a successful (200 OK) HttpResponse,
      // but has an invalid response body format.
      // if (checkInvalid200Response(httpEvent)) {
      //   // Throws a custom HttpResponseBodyFormatError to signal the invalid response format.
      //   // This error will be caught and handled by the subsequent catchError operator,
      //   // which will display the appropriate error message to the user.
      //   throw new HttpResponseBodyFormatError();
      // }
    }),
    catchError((error) => {
      // let errorMessage: string;

      // Handle network connection errors specifically
      // if (checkNoNetworkConnection(error)) {
      //   // Set specific message for network errors
      //   errorMessage = MESSAGES.NO_CONNECTION;

      //   // Create a custom network error object
      //   error = new HttpNoNetworkConnectionError();

      //   // Mark the error as caught to prevent duplicate handling
      //   error.wasCaught = true;
      // }
      // else if (is400ResponseError(error)) {
      //   // Explicitly skip handling 400 errors here (handled by tapValidationErrors operator)
      //   // This ensures that validation errors are handled in the component,
      //   // while other errors (e.g., 5xx, 4xx) fall through to the next case.
      //   errorMessage = '';
      // }
      // else {
      //   // For all other server errors or unexpected errors, display a generic error message.
      //   errorMessage = MESSAGES.INTERNAL_ERROR
      // }

      // // Show a Snackbar notification if an error message is available.
      // if (errorMessage) {
      //   snackbar.open(errorMessage);
      // }

      if (checkUnauthorizedResponse(error)) {
        // redirect to login page
        document.location.href = '/login';
      }

      // Re-throw the error for handling in the component or a global error handler.
      throw error;
    })
  );
};

function checkUnauthorizedResponse(error: any) {
  return error instanceof HttpErrorResponse && error.status === 401;
}
