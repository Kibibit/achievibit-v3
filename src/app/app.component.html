<header>
  <nav>
    <button class="circle transparent">
      <i>menu</i>
    </button>
    <h5 class="max center-align">achievibit</h5>
    <button class="circle transparent" *ngIf="loggedInUser">
      <img class="responsive" [src]="loggedInUser.avatar">
    </button>
  </nav>
</header>
<main class="main">
  <div class="content">
    <div>
      <div class="tabs">
        <a class="active">GitHub</a>
        <a>GitLab</a>
        <a>BitBucket</a>
      </div>
      <div class="page padding active">
        <h5>GitHub Repos</h5>
        <div class="kb-repo" *ngFor="let repo of userRepos" [class.private]="repo.private"
          [class.public]="!repo.private" [class.github]="repo.system === 'github'"
          [class.bitbucket]="repo.system === 'bitbucket'" [class.gitlab]="repo.system === 'gitlab'">
          <div class="kb-repo-name">
            {{ repo.name }}
            <i class="visibility-icon"></i>
          </div>
          <div>{{ repo.full_name }}</div>
        </div>
      </div>
      <div class="page padding">
        <h5>Tab 2</h5>
      </div>
      <div class="page padding">
        <h5>Tab 3</h5>
      </div>
    </div>
  </div>
</main>

<!-- * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * -->
<!-- * * * * * * * * * * * The content above * * * * * * * * * * * * -->
<!-- * * * * * * * * * * is only a placeholder * * * * * * * * * * * -->
<!-- * * * * * * * * * * and can be replaced.  * * * * * * * * * * * -->
<!-- * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * -->
<!-- * * * * * * * * * * End of Placeholder  * * * * * * * * * * * * -->
<!-- * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * -->


<router-outlet />