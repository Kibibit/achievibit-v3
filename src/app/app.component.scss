.logo {
  height: 30px;
  margin-right: 0.5em;
}

.menu {
  margin: 1em;
}

main {
  position: relative;
}