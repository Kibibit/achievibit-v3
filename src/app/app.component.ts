import { combineLatest } from 'rxjs';
import { NgFor, NgIf } from '@angular/common';
import { Component, OnInit } from '@angular/core';
import { RouterLink, RouterOutlet } from '@angular/router';

import { AppService } from './app.service';

@Component({
  selector: 'kb-root',
  standalone: true,
  imports: [ RouterOutlet, NgIf, NgFor, RouterLink ],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss'
})
export class AppComponent implements OnInit {
  loggedInUser: any;

  constructor(
    private appService: AppService
  ) {}

  ngOnInit(): void {
    const loggedInUserObs = this
      .appService
      .getLoggedInUser();

    // Combine the two observables to get the user and their repos
    combineLatest([ loggedInUserObs ])
      .subscribe(([ user ]) => {
        this.loggedInUser = user;
      });
  }
}
