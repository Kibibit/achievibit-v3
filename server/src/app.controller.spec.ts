import { Test, TestingModule } from '@nestjs/testing';

import { AppController } from './app.controller';
import { AppService } from './app.service';
import { SmeeService } from '@kb-config';

jest.mock('fs-extra', () => ({
  readJSON: jest.fn().mockResolvedValue({
    name: 'test',
    description: 'test',
    version: 'test',
    license: 'test',
    repository: 'test',
  })
}));

describe('AppController', () => {
  let appController: AppController;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      controllers: [ AppController ],
      providers: [
        AppService,
        SmeeService
      ]
    }).compile();

    appController = app.get<AppController>(AppController);
  });

  describe('root', () => {
    it('should return "Hello World!"', async () => {
      expect(appController.getApiDetails()).resolves.toMatchSnapshot();
    });
  });
});
