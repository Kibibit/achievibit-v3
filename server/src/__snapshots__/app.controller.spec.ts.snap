// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController root should return "Hello World!" 1`] = `
ApiInfo {
  "description": "test",
  "license": "test",
  "name": "test",
  "repository": "test",
  "version": "test",
}
`;
